core/src/main/scala/kafka/server/DynamicBrokerConfig.scala

```diff
@@ -343,18 +355,44 @@ class DynamicBrokerConfig(private val kafkaConfig: KafkaConfig) extends Logging
       currentConfig
   }
 
-  private def processReconfigurable(reconfigurable: Reconfigurable, updatedKeys: Set[String],
-                                    allNewConfigs: util.Map[String, _], newCustomConfigs: util.Map[String, Object],
+  private def needsReconfiguration(reconfigurable: Reconfigurable, updatedKeys: Set[String]): Boolean = {
+    reconfigurable.reconfigurableConfigs.asScala.intersect(updatedKeys).nonEmpty
+  }
+
+  private def processReconfigurable(reconfigurable: Reconfigurable,
+                                    allNewConfigs: util.Map[String, _],
+                                    newCustomConfigs: util.Map[String, Object],
                                     validateOnly: Boolean): Unit = {
-    if (reconfigurable.reconfigurableConfigs.asScala.intersect(updatedKeys).nonEmpty) {
-      val newConfigs = new util.HashMap[String, Object]
-      allNewConfigs.asScala.foreach { case (k, v) => newConfigs.put(k, v.asInstanceOf[AnyRef]) }
-      newConfigs.putAll(newCustomConfigs)
-      if (validateOnly) {
-        if (!reconfigurable.validateReconfiguration(newConfigs))
-          throw new ConfigException("Validation of dynamic config update failed")
-      } else
-        reconfigurable.reconfigure(newConfigs)
-    }
+    val newConfigs = new util.HashMap[String, Object]
+    allNewConfigs.asScala.foreach { case (k, v) => newConfigs.put(k, v.asInstanceOf[AnyRef]) }
+    newConfigs.putAll(newCustomConfigs)
+    if (validateOnly) {
+      if (!reconfigurable.validateReconfiguration(newConfigs))
+        throw new ConfigException("Validation of dynamic config update failed")
+    } else
+      reconfigurable.reconfigure(newConfigs)
+  }
+
+  private def processBrokerReconfigurable(reconfigurable: BrokerReconfigurable,
+                                    oldConfig: KafkaConfig,
+                                    newConfig: KafkaConfig,
+                                    validateOnly: Boolean): Unit = {
+    if (validateOnly) {
+      if (!reconfigurable.validateBrokerReconfiguration(newConfig))
+        throw new ConfigException("Validation of dynamic config update failed")
+    } else
+      reconfigurable.brokerReconfigure(oldConfig, newConfig)
   }
 }
+
+trait BrokerReconfigurable extends Reconfigurable {
```

Comment:
The overrides below make me a little uncomfortable. I guess the only reason we need this to extend `Reconfigurable` is because of the `reconfigurables` collection? Do you think it would be unreasonable to use two separate collections?

Reply:
@hachikuji Makes sense, updated. Thanks.